### PR TITLE
docs: add package godoc for indexer

### PR DESCRIPTION
**Pull request by alexanderbez**

### CODE REVIEW

File: state/indexer/doc.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+Package indexer defines Tendermint's block and transaction event indexing logic.
+
+Currently, Tendermint supports two primary means of block and transaction event
+indexing:
```
Comment (creachadair) [inline state/indexer/doc.go:5]:
```suggestion
Tendermint supports two primary means of block and transaction event indexing:
```

File: state/indexer/doc.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+1. A key-value sink via an embedded database with a proprietary query language.
+2. A Postgres-based sink.
+
+Recall, an ABCI application can emit events during block and transaction execution
+in the form of <abci.Event.Type>.<abci.EventAttributeKey>=<abci.EventAttributeValue>,
+e.g. 'transfer.amount=10000'.
```
Comment (creachadair) [inline state/indexer/doc.go:12]:
```suggestion
An ABCI application can emit events during block and transaction execution in the form

   <abci.Event.Type>.<abci.EventAttributeKey>=<abci.EventAttributeValue>

for example "transfer.amount=10000".
```

File: state/indexer/doc.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+limited, but does allow users to directly query for block and transaction events
+against Tendermint's RPC. Instead, operators are encouraged to use the "psql"
+indexing sink when more complex queries are required and for reliability purposes
+as PostgreSQL can scale.
+
+Prior to starting Tendermint with the "psql" indexing sink enabled, operators
+must ensure the following:
+
+1. The "psql" indexing sink is provided in Tendermint's configuration.
+2. A 'tx-index.psql-conn' value is provided that contains the PostgreSQL connection URI.
```
Comment (creachadair) [inline state/indexer/doc.go:40]: I'm not sure if there's a rule behind when you're using 'single quotes' vs. "double quotes" but I'd suggest sticking to just one of them (personally I like double, but I'm not picky).

File: state/indexer/doc.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+Example:
+
+	[tx-index]
+	indexer = ["kv", "psql"]
+
+If an operator wants to completely disable indexing, they may simply just provide
+the "null" sink option in the configuration. All other sinks will be ignored if
+"null" is provided.
+
+If indexing is enabled, the indexer.Service will iterate over all enabled sinks
```
Comment (creachadair) [inline state/indexer/doc.go:26]:
```suggestion
If indexing is enabled, the indexer service will iterate over all enabled sinks
```

or perhaps "indexer.Service", but I don't know how necessary it is to spell out the type vs. just describing it. YMMV.

File: state/indexer/doc.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+and invoke block and transaction indexing via the appropriate IndexBlockEvents
+and IndexTxEvents methods.
+
+Note, the "kv" sink is considered deprecated and it's query functionality is very
+limited, but does allow users to directly query for block and transaction events
+against Tendermint's RPC. Instead, operators are encouraged to use the "psql"
+indexing sink when more complex queries are required and for reliability purposes
+as PostgreSQL can scale.
```
Comment (creachadair) [inline state/indexer/doc.go:34]: N.B. typo on "it's" even if the other suggestions here are not to your liking.

```suggestion
Note, the "kv" sink is deprecated and its query functionality is very limited. It should
be used only for testing or setup, not in production. Instead, operators are encouraged
to use the PostgreSQL ("psql") indexing sink when more complex queries are required
and for greater reliability.
```

I suggested dropping the mention of querying via the RPC, on the assumption that if kv is deprecated and psql doesn't support the RPC query, the RPC query should probably be deprecated too. But if that's not true, maybe we should reconsider what we're telling people. IMO we shouldn't give pride of place to deprecated things in the package docs.

File: state/indexer/doc.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+1. The "psql" indexing sink is provided in Tendermint's configuration.
+2. A 'tx-index.psql-conn' value is provided that contains the PostgreSQL connection URI.
+3. The block and transaction event schemas have been created in the PostgreSQL database.
+
+Tendermint provides the block and transaction event schemas in the following
+path: state/indexer/sink/psql/schema.sql
+
+To create the schema in a PostgreSQL database, operators can perform the query
+with the schema manually or invoke schema creation via the CLI:
```
Comment (creachadair) [inline state/indexer/doc.go:47]:
```suggestion
To create the schema in a PostgreSQL database, perform the schema query
manually or invoke schema creation via the CLI:
```

File: state/indexer/doc.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+	$ psql <flags> -f state/indexer/sink/psql/schema.sql
+
+Note, using the "psql" indexing sink prohibits queries against the RPC. However,
+this is due to the fact that querying can and should be done directly against
+the PostgreSQL database instead as it offers a rich query language and features.
```
Comment (creachadair) [inline state/indexer/doc.go:53]:
```suggestion
The "psql" indexing sink prohibits queries via RPC. When using a PostgreSQL sink,
queries can and should be made directly against the database using SQL.
```

File: state/indexer/doc.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+Operators are able to perform queries such as the following:
```
Comment (creachadair) [inline state/indexer/doc.go:55]:
```suggestion
The following are some example SQL queries against the database schema:
```

File: state/indexer/doc.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	SELECT * FROM tx_events WHERE height = 25;
+
+* Query for transaction events that have a given type (i.e. value wildcard):
+
+	SELECT * FROM tx_events WHERE key LIKE '%transfer.recipient%';
+
+
+Note, if the entire abci.TxResult is needed, a foreign key exists in the tx_events
+table that maps to tx_results. The tx_results table contains the raw Protobuf
+encoded abci.TxResult object.
```
Comment (creachadair) [inline state/indexer/doc.go:72]:
```suggestion
Note that if a complete abci.TxResult is needed, you will need to join "tx_events" with
"tx_results" via a foreign key, to obtain contains the raw protobuf-encoded abci.TxResult.
```

**Review by creachadair — COMMENTED** [review]
This is really helpful, thank you. I made some suggestions to tighten up the wording in a few places, and to fix a couple of typos, but all of these should be seen as optional stylistic things (and if my suggestions introduced anything wrong, of course they should be ignored).

File: state/indexer/doc.go (hunk 4 above)
Comment (alexanderbez) [reply state/indexer/doc.go:27]: No need for this change.

**Review by creachadair — APPROVED** [review]
There are a couple of unresolved comments I'd like to see addressed or resolved before this merges, but otherwise this LGTM.

N.B. Even if you don't apply my suggestion at Line 35, please fix the typo on "it's".

**Pull request merged**